File: Source/Relativity.DataExchange.Core/Transfer/LongTextStreamService.cs
// --------------------------------------------------------------------------------------------------------------------
// <copyright file="LongTextStreamService.cs" company="Relativity ODA LLC">
//   © Relativity All Rights Reserved.
// </copyright>
// <summary>
//   Represents a service class object to retrieve long text data through the Object Manager streaming API.
// </summary>
// --------------------------------------------------------------------------------------------------------------------

namespace Relativity.DataExchange.Transfer
{
	using System;
	using System.IO;
	using System.Threading;
	using System.Threading.Tasks;

	using Polly;

	using Relativity.DataExchange.Io;
	using Relativity.DataExchange.Service;
	using Relativity.Kepler.Transport;
	using Relativity.Logging;
	using Relativity.Services.Objects;
	using Relativity.Services.Objects.DataContracts;

	/// <summary>
	/// Represents a service class object to retrieve long text data through the Object Manager streaming API.
	/// </summary>
	internal class LongTextStreamService : ILongTextStreamService
	{
		private const string RetryCountKey = "RetryCount";
		private const string RequestKey = "Request";
		private readonly IServiceProxyFactory serviceProxyFactory;
		private readonly IAppSettings settings;
		private readonly IFileSystem fileSystem;
		private readonly ILog logger;
		private readonly int exportLongTextBufferSizeBytes;
		private readonly int exportLongTextLargeFileProgressRateSeconds;
		private readonly ISyncPolicy syncPolicy;
		private readonly IAsyncPolicy asyncRetryPolicy;
		private readonly Lazy<IObjectManager> objectManager;
		private bool disposed;

		/// <summary>
		/// Initializes a new instance of the <see cref="LongTextStreamService"/> class.
		/// </summary>
		/// <param name="serviceProxyFactory">
		/// The factory used to create service proxy instances.
		/// </param>
		/// <param name="serviceRetryPolicyFactory">
		/// The factory used to create service retry policies.
		/// </param>
		/// <param name="settings">
		/// The settings.
		/// </param>
		/// <param name="fileSystem">
		/// The file system wrapper.
		/// </param>
		/// <param name="logger">
		/// The logger instance.
		/// </param>
		public LongTextStreamService(
			IServiceProxyFactory serviceProxyFactory,
			IServiceRetryPolicyFactory serviceRetryPolicyFactory,
			IAppSettings settings,
			IFileSystem fileSystem,
			ILog logger)
		{
			this.serviceProxyFactory = serviceProxyFactory.ThrowIfNull(nameof(serviceProxyFactory));
			this.settings = settings.ThrowIfNull(nameof(settings));
			this.fileSystem = fileSystem.ThrowIfNull(nameof(fileSystem));
			this.logger = logger.ThrowIfNull(nameof(logger));
			this.exportLongTextBufferSizeBytes = settings.ExportLongTextBufferSizeBytes;
			this.exportLongTextLargeFileProgressRateSeconds = settings.ExportLongTextLargeFileProgressRateSeconds;
			this.objectManager = new Lazy<IObjectManager>(this.CreateObjectManager);
			serviceRetryPolicyFactory.ThrowIfNull(nameof(serviceRetryPolicyFactory));
			this.syncPolicy = serviceRetryPolicyFactory.CreatePolicy(this.OnRetryCreateObjectManagerInstance);
			this.asyncRetryPolicy = serviceRetryPolicyFactory.CreateAsyncPolicy(this.OnRetrySaveLongTextStream);
		}

		/// <inheritdoc />
		public void Dispose()
		{
			this.Dispose(true);
		}

		/// <inheritdoc />
		public Task<LongTextStreamResult> SaveLongTextStreamAsync(
			LongTextStreamRequest request,
			CancellationToken token,
			IProgress<LongTextStreamProgressEventArgs> progress)
		{
			request.ThrowIfNull(nameof(request));
			ValidateRequest(request);
			Context context = new Context("LongTextExecutionKey") { { RetryCountKey, 0 } };
			return this.asyncRetryPolicy.ExecuteAsync(
				(ctx, ct) => this.OnExecuteSaveLongTextStream(request, context, token, progress),
				context,
				token);
		}

		private static int GetRetryCount(Context context)
		{
			return context.TryGetValue(RetryCountKey, out var value) ? Convert.ToInt32(value) : 0;
		}

		private static LongTextStreamRequest GetLongTextStreamRequest(Context context)
		{
			return context.TryGetValue(RequestKey, out var value)
				       ? value as LongTextStreamRequest
				       : new LongTextStreamRequest();
		}

		private static void ValidateRequest(LongTextStreamRequest request)
		{
			if (request.SourceObjectArtifactId < 1 || request.SourceFieldArtifactId < 1)
			{
				throw new ArgumentException(
					$"The long text streaming service request requires a source {(request.SourceFieldArtifactId < 1 ? "field" : "object")} artifact identifier greater than zero.",
					nameof(request));
			}

			if (string.IsNullOrWhiteSpace(request.TargetFile))
			{
				throw new ArgumentException(
					"The long text streaming service request requires the target file to be non-empty.",
					nameof(request));
			}

			if (request.TargetEncoding == null || request.SourceEncoding == null)
			{
				throw new ArgumentException(
					$"The long text streaming service request requires the {(request.SourceEncoding == null ? "source" : "target")} encoding to be defined.",
					nameof(request));
			}

			if (request.WorkspaceId < 1)
			{
				throw new ArgumentException(
					$"The long text streaming service request requires the workspace artifact identifier to be greater than zero.",
					nameof(request));
			}
		}

		private IObjectManager CreateObjectManager()
		{
			// This will address potential issues with exception caching.
			return this.syncPolicy.Execute(() => this.serviceProxyFactory.CreateProxyInstance<IObjectManager>());
		}

		private void LogNonFatalInvalidRequestWarning(LongTextStreamRequest request, Exception exception)
		{
			this.logger.LogWarning(
				exception,
				"The {ArtifactId} - {FieldArtifactId} long text request returned a non-fatal and non-retryable error because 1 or more parameters is invalid.",
				request.SourceObjectArtifactId,
				request.SourceFieldArtifactId);
		}

		private void LogNonFatalObjectManagerWarning(LongTextStreamRequest request, Exception exception)
		{
			this.logger.LogWarning(
				exception,
				"The {ArtifactId} - {FieldArtifactId} long text request returned a non-fatal and non-retryable object manager specific error.",
				request.SourceObjectArtifactId,
				request.SourceFieldArtifactId);
		}

		private async Task<LongTextStreamResult> OnExecuteSaveLongTextStream(LongTextStreamRequest request, Context context, CancellationToken token, IProgress<LongTextStreamProgressEventArgs> progress)
		{
			int retryCount = GetRetryCount(context);

			try
			{
				context[RequestKey] = request;
				RelativityObjectRef exportObject =
					new RelativityObjectRef { ArtifactID = request.SourceObjectArtifactId };
				FieldRef longTextField = new FieldRef { ArtifactID = request.SourceFieldArtifactId };
				try
				{
					using (IKeplerStream keplerStream = await this.objectManager.Value
						                                    .StreamLongTextAsync(
							                                    request.WorkspaceId,
							                                    exportObject,
							                                    longTextField).ConfigureAwait(false))
					using (Stream sourceStream = await keplerStream.GetStreamAsync().ConfigureAwait(false))
					using (FileStream targetStream = this.fileSystem.File.Create(request.TargetFile))
					{
						return this.CopySourceStreamData(request, context, sourceStream, targetStream, token, progress);
					}
				}
				catch (Exception)
				{
					this.TryDeleteTargetFile(request.TargetFile);
					throw;
				}
			}
			catch (ArgumentException e)
			{
				// Non-OM invalid request or path.
				this.LogNonFatalInvalidRequestWarning(request, e);
				return LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e);
			}
			catch (NotSupportedException e)
			{
				// Non-OM invalid path.
				this.LogNonFatalInvalidRequestWarning(request, e);
				return LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e);
			}
			catch (Exception e)
			{
				if (ObjectManagerExceptionHelper.IsNonFatalError(e))
				{
					this.LogNonFatalObjectManagerWarning(request, e);
					return LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e);
				}

				// All other errors are logged and will go through the service retry policy.
				throw;
			}
		}

		private void OnRetryCreateObjectManagerInstance(Exception exception, TimeSpan duration, int retryCount, Context context)
		{
			this.logger.LogError(
				exception,
				"Failed to create the ObjectManager proxy instance. Currently on attempt {RetryCount} out of {MaxRetries} and waiting {WaitSeconds} seconds before the next retry attempt.",
				retryCount,
				this.settings.HttpErrorNumberOfRetries,
				duration.TotalSeconds);
		}

		private void OnRetrySaveLongTextStream(Exception exception, TimeSpan duration, int retryCount, Context context)
		{
			context[RetryCountKey] = retryCount;
			LongTextStreamRequest request = GetLongTextStreamRequest(context);
			this.logger.LogError(
				exception,
				"An error occurred retrieving the {ArtifactId} - {FieldArtifactId} long text data from Object Manager. Currently on attempt {RetryCount} out of {MaxRetries} and waiting {WaitSeconds} seconds before the next retry attempt.",
				request.SourceObjectArtifactId,
				request.SourceFieldArtifactId,
				retryCount,
				this.settings.HttpErrorNumberOfRetries,
				duration.TotalSeconds);
		}

		private LongTextStreamResult CopySourceStreamData(
			LongTextStreamRequest request,
			Context context,
			Stream sourceStream,
			FileStream targetStream,
			CancellationToken token,
			IProgress<LongTextStreamProgressEventArgs> progress)
		{
			// Note: this method is critical - avoid all possible overhead (e.g. Stopwatch).
			// Note: synchronous read/write operations are significantly faster than their asynchronous counterparts.
			long totalBytesWritten = 0;
			DateTime startTimestamp = DateTime.Now;
			const bool DetectEncodingFromByteOrderMarks = false;
			using (StreamReader reader = new StreamReader(
				sourceStream,
				request.SourceEncoding,
				DetectEncodingFromByteOrderMarks,
				this.exportLongTextBufferSizeBytes))
			using (StreamWriter writer = new StreamWriter(
				targetStream,
				request.TargetEncoding,
				this.exportLongTextBufferSizeBytes))
			{
				char[] buffer = new char[this.exportLongTextBufferSizeBytes / sizeof(char)];
				DateTime progressTimestamp = startTimestamp;
				bool continueReading = true;
				while (continueReading)
				{
					token.ThrowIfCancellationRequested();
					try
					{
						int chunkCharacters = reader.Read(buffer, 0, buffer.Length);
						if (chunkCharacters == 0)
						{
							continueReading = false;
							writer.Flush();
						}
						else
						{
							writer.Write(buffer, 0, chunkCharacters);
						}
					}
					finally
					{
						totalBytesWritten = writer.BaseStream.Length;
						if ((!continueReading && startTimestamp != progressTimestamp)
						    || (DateTime.Now - progressTimestamp).TotalSeconds
						    >= this.exportLongTextLargeFileProgressRateSeconds)
						{
							progress?.Report(
								new LongTextStreamProgressEventArgs(request, totalBytesWritten, !continueReading));
							progressTimestamp = DateTime.Now;
						}
					}
				}

				DateTime endTimestamp = DateTime.Now;
				return LongTextStreamResult.CreateSuccessfulResult(
					request,
					request.TargetFile,
					totalBytesWritten,
					GetRetryCount(context),
					endTimestamp - startTimestamp);
			}
		}

		private void Dispose(bool disposing)
		{
			if (disposing && !this.disposed)
			{
				if (this.objectManager.IsValueCreated)
				{
					this.objectManager.Value.Dispose();
				}

				this.serviceProxyFactory?.Dispose();
				this.disposed = true;
			}
		}

		[System.Diagnostics.CodeAnalysis.SuppressMessage(
			"Microsoft.Design",
			"CA1031:DoNotCatchGeneralExceptionTypes",
			Justification = "This is OK for what this method is doing and prefixed the method name to make it obvious.")]
		private void TryDeleteTargetFile(string file)
		{
			try
			{
				// Don't allow expected errors to rethrow.
				this.fileSystem.File.Delete(file);
			}
			catch (Exception e)
			{
				this.logger.LogWarning(e, "Failed to delete the target file.");
			}
		}
	}
}

File: Source/Relativity.DataExchange.Legacy/Service/Kepler/LongTextStreamService.vb
' --------------------------------------------------------------------------------------------------------------------
' <copyright file="LongTextStreamService.cs" company="Relativity ODA LLC">
'   © Relativity All Rights Reserved.
' </copyright>
' <summary>
'   Represents a service class object to retrieve long text data through the Object Manager streaming API.
' </summary>
' --------------------------------------------------------------------------------------------------------------------

Imports System
Imports System.IO
Imports System.Threading
Imports System.Threading.Tasks
Imports kCura.WinEDDS.Service.Replacement
Imports Polly
Imports Relativity.DataExchange
Imports Relativity.DataExchange.Io
Imports Relativity.DataExchange.Service
Imports Relativity.DataExchange.Transfer
Imports Relativity.Kepler.Transport
Imports Relativity.Logging
Imports Relativity.Services.Objects
Imports Relativity.Services.Objects.DataContracts
Imports IObjectManager = Relativity.Services.Objects.IObjectManager

Namespace kCura.WinEDDS.Service.Kepler

	''' <summary>
	''' Represents a service class object to retrieve long text data through the Object Manager streaming API.
	''' </summary>
	Friend Class LongTextStreamService
		Implements ILongTextStreamService

		Private Const LongTextRetryCountKey As String = "LongTextRetryCount"
		Private Const LongTextRequestKey As String = "LongTextRequest"
		Private ReadOnly serviceNotification As IServiceNotification
		Private ReadOnly settings As IAppSettings
		Private ReadOnly fileSystem As IFileSystem
		Private ReadOnly logger As ILog
		Private ReadOnly exportLongTextBufferSizeBytes As Integer
		Private ReadOnly exportLongTextLargeFileProgressRateSeconds As Integer
		Private ReadOnly keplerProxy As IKeplerProxy

		Public Sub New(ByVal serviceProxyFactory As IServiceProxyFactory, ByVal serviceNotification As IServiceNotification, ByVal settings As IAppSettings, ByVal fileSystem As IFileSystem, ByVal logger As ILog)
			Me.serviceNotification = serviceNotification.ThrowIfNull(NameOf(serviceNotification))
			Me.settings = settings.ThrowIfNull(NameOf(settings))
			Me.fileSystem = fileSystem.ThrowIfNull(NameOf(fileSystem))
			Me.logger = logger.ThrowIfNull(NameOf(logger))
			serviceProxyFactory.ThrowIfNull(NameOf(serviceProxyFactory))

			Me.keplerProxy = new KeplerProxy(serviceProxyFactory, logger, Sub(exception, duration, retryCount, context) OnRetrySaveLongTextStream(exception, duration, retryCount, context))
			exportLongTextBufferSizeBytes = settings.ExportLongTextBufferSizeBytes
			exportLongTextLargeFileProgressRateSeconds = settings.ExportLongTextLargeFileProgressRateSeconds
		End Sub

		''' <inheritdoc/>
		Public Sub Dispose() Implements IDisposable.Dispose
		End Sub

		''' <inheritdoc/>
		Public Function SaveLongTextStreamAsync(ByVal request As Transfer.LongTextStreamRequest, ByVal token As CancellationToken, ByVal progress As IProgress(Of Transfer.LongTextStreamProgressEventArgs)) As Task(Of Transfer.LongTextStreamResult) Implements Transfer.ILongTextStreamService.SaveLongTextStreamAsync
			request.ThrowIfNull(NameOf(request))
			LongTextStreamService.ValidateRequest(request)
			Dim context As Context = New Context("LongTextExecutionKey") From {
				{LongTextRequestKey, request},
				{LongTextRetryCountKey, 0}
			}
			Return keplerProxy.ExecuteAsync(context, token, Function(ctx, ct, serviceProxyFactory) Me.OnExecuteSaveLongTextStream(serviceProxyFactory, request, context, token, progress))
		End Function

		Private Shared Function GetRetryCount(ByVal context As Context) As Integer
			Dim value As Object = Nothing
			Return If(context.TryGetValue(LongTextRetryCountKey, value), Convert.ToInt32(value), 0)
		End Function

		Private Shared Function GetLongTextStreamRequest(ByVal context As Context) As Transfer.LongTextStreamRequest
			Dim value As Object = Nothing
			Return If(context.TryGetValue(LongTextRequestKey, value), TryCast(value, Transfer.LongTextStreamRequest), New Transfer.LongTextStreamRequest())
		End Function

		Private Shared Sub ValidateRequest(ByVal request As Transfer.LongTextStreamRequest)
			If request.SourceObjectArtifactId < 1 OrElse request.SourceFieldArtifactId < 1 Then
				Throw New ArgumentException($"The long text streaming service request requires a source {If(request.SourceFieldArtifactId < 1, "field", "object")} artifact identifier greater than zero.", NameOf(request))
			End If

			If String.IsNullOrWhiteSpace(request.TargetFile) Then
				Throw New ArgumentException("The long text streaming service request requires the target file to be non-empty.", NameOf(request))
			End If

			If request.TargetEncoding Is Nothing OrElse request.SourceEncoding Is Nothing Then
				Throw New ArgumentException($"The long text streaming service request requires the {If(request.SourceEncoding Is Nothing, "source", "target")} encoding to be defined.", NameOf(request))
			End If

			If request.WorkspaceId < 1 Then
				Throw New ArgumentException($"The long text streaming service request requires the workspace artifact identifier to be greater than zero.", NameOf(request))
			End If
		End Sub

		Private Sub LogNonFatalInvalidRequestWarning(ByVal request As Transfer.LongTextStreamRequest, ByVal exception As Exception)
			logger.LogWarning(exception, "The {ArtifactId} - {FieldArtifactId} long text request returned a non-fatal and non-retryable error because 1 or more parameters is invalid.", request.SourceObjectArtifactId, request.SourceFieldArtifactId)
		End Sub

		Private Sub LogNonFatalObjectManagerWarning(ByVal request As Transfer.LongTextStreamRequest, ByVal exception As Exception)
			logger.LogWarning(exception, "The {ArtifactId} - {FieldArtifactId} long text request returned a non-fatal and non-retryable object manager specific error.", request.SourceObjectArtifactId, request.SourceFieldArtifactId)
		End Sub

		Private Async Function OnExecuteSaveLongTextStream(ByVal serviceProxyFactory As IServiceProxyFactory, ByVal request As Transfer.LongTextStreamRequest, ByVal context As Context, ByVal token As CancellationToken, ByVal progress As IProgress(Of Transfer.LongTextStreamProgressEventArgs)) As Task(Of Transfer.LongTextStreamResult)
			Dim retryCount As Integer = GetRetryCount(context)

			Try
				Dim exportObject As RelativityObjectRef = New RelativityObjectRef With {
					.ArtifactID = request.SourceObjectArtifactId
				}
				Dim longTextField As FieldRef = New FieldRef With {
					.ArtifactID = request.SourceFieldArtifactId
				}

				Try

					Using objectManager As IObjectManager = serviceProxyFactory.CreateProxyInstance(Of IObjectManager)()

						Using keplerStream As IKeplerStream = Await objectManager.StreamLongTextAsync(CInt(request.WorkspaceId), CType(exportObject, RelativityObjectRef), CType(longTextField, FieldRef)).ConfigureAwait(False)

							Using sourceStream As Stream = Await keplerStream.GetStreamAsync().ConfigureAwait(False)

								Using targetStream As FileStream = fileSystem.File.Create(request.TargetFile)
									Return Me.CopySourceStreamData(request, context, sourceStream, targetStream, token, progress)
								End Using
							End Using
						End Using
					End Using

				Catch __unusedException1__ As Exception
					Me.TryDeleteTargetFile(request.TargetFile)
					Throw
				End Try

			Catch e As ArgumentException
				' Non-OM invalid request or path.
				Me.LogNonFatalInvalidRequestWarning(request, e)
				Return Transfer.LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e)
			Catch e As NotSupportedException
				' Non-OM invalid path.
				Me.LogNonFatalInvalidRequestWarning(request, e)
				Return Transfer.LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e)
			Catch e As Exception

				If ObjectManagerExceptionHelper.IsNonFatalError(e) Then
					Me.LogNonFatalObjectManagerWarning(request, e)
					Return Transfer.LongTextStreamResult.CreateNonFatalIssueResult(request, retryCount, e)
				End If


				' All other errors are logged and will go through the service retry policy.
				Throw
			End Try
		End Function

		Private Sub OnRetrySaveLongTextStream(ByVal exception As Exception, ByVal duration As TimeSpan, ByVal retryCount As Integer, ByVal context As Context)
			' Intentionally limit notifications to status messages to avoid a large numbers of warning messages and total warning counts.
			context(LongTextRetryCountKey) = retryCount
			Dim request As Transfer.LongTextStreamRequest = GetLongTextStreamRequest(context)
			logger.LogError(exception, "An error occurred retrieving the {ArtifactId} - {FieldArtifactId} long text data from Object Manager. Currently on attempt {RetryCount} out of {MaxRetries} and waiting {WaitSeconds} seconds before the next retry attempt.", request.SourceObjectArtifactId, request.SourceFieldArtifactId, retryCount, settings.HttpErrorNumberOfRetries, duration.TotalSeconds)
			serviceNotification.NotifyStatus(ErrorMessageFormatter.FormatWebServiceRetryMessage("Long Text Stream Request", exception.Message, duration, retryCount, settings.HttpErrorNumberOfRetries))
		End Sub

		Private Function CopySourceStreamData(ByVal request As Transfer.LongTextStreamRequest, ByVal context As Context, ByVal sourceStream As Stream, ByVal targetStream As FileStream, ByVal token As CancellationToken, ByVal progress As IProgress(Of Transfer.LongTextStreamProgressEventArgs)) As Transfer.LongTextStreamResult
			' Note: this method is critical - avoid all possible overhead (e.g. Stopwatch).
			' Note: synchronous read/write operations are significantly faster than their asynchronous counterparts.
			Dim totalBytesWritten As Long = 0
			Dim startTimestamp As Date = Date.Now
			Const DetectEncodingFromByteOrderMarks As Boolean = False

			Using reader As StreamReader = New StreamReader(sourceStream, request.SourceEncoding, DetectEncodingFromByteOrderMarks, exportLongTextBufferSizeBytes)

				Using writer As StreamWriter = New StreamWriter(targetStream, request.TargetEncoding, exportLongTextBufferSizeBytes)
					Dim buffer As Char() = New Char(CInt(Me.exportLongTextBufferSizeBytes / Len(New Char) - 1)) {}
					Dim progressTimestamp As Date = startTimestamp
					Dim continueReading As Boolean = True

					While continueReading
						token.ThrowIfCancellationRequested()

						Try
							Dim chunkCharacters As Integer = reader.Read(buffer, 0, buffer.Length)

							If chunkCharacters = 0 Then
								continueReading = False
								writer.Flush()
							Else
								writer.Write(buffer, 0, chunkCharacters)
							End If

						Finally
							totalBytesWritten = writer.BaseStream.Length

							If Not continueReading AndAlso startTimestamp <> progressTimestamp OrElse (Date.Now - progressTimestamp).TotalSeconds >= exportLongTextLargeFileProgressRateSeconds Then
								progress?.Report(New Transfer.LongTextStreamProgressEventArgs(request, totalBytesWritten, Not continueReading))
								progressTimestamp = Date.Now
							End If
						End Try
					End While

					Dim endTimestamp As Date = Date.Now
					Return Transfer.LongTextStreamResult.CreateSuccessfulResult(request, request.TargetFile, totalBytesWritten, GetRetryCount(context), endTimestamp - startTimestamp)
				End Using
			End Using
		End Function

		Private Sub TryDeleteTargetFile(ByVal file As String)
			Try
				' Don't allow expected errors to rethrow.
				fileSystem.File.Delete(file)
			Catch e As Exception
				logger.LogWarning(e, "Failed to delete the target file.")
			End Try
		End Sub
	End Class
End Namespace